chain/client/src/client_actor.rs

```diff
@@ -378,17 +380,20 @@ impl Handler<NetworkClientMessages> for ClientActor {
                             }
                         }
                         ShardSyncStatus::StateDownloadParts => {
-                            for part in shard_state.parts.iter() {
-                                let part_id = part.state_part.part_id as usize;
-                                if part_id >= shard_sync_download.downloads.len() {
-                                    // TODO ???
+                            for (i, part_id) in shard_state.part_ids.iter().enumerate() {
+                                let part_id = *part_id as usize;
+                                let num_parts = shard_sync_download.downloads.len();
+                                if part_id >= num_parts {
+                                    // TODO MOO ask Alex if continue is enough here
```

Comment:
seems like TODO is to ban the peer

Reply:
Example.
1. Node 1 asks for the state and requests a header.
2. Node 2 (malicious) returns huge state size.
3. Node 1 decides to divide the state into 100 parts.
4. Node 1 requests part 99 from Node 3.
5. Node 1 collect the state but cannot finalize it. Node 1 starts state sync from the beginning.
6. Node 1 requests header from Node 4. Size of the state is small, and Node 1 decides to divide the state into 5 parts.
7. Node 3 returns a part with part_id 99.

I don't think we should ban Node 3 in this case. BTW, I'm not sure if my case is possible.

Reply:
I don't think the scenario is possible unless someone gave you a wrong block header, right?
As in, the state DL header is fully validated against the block header, and unless the block header is already forged, the state DL header wouldn't pass the checks if it had anything funky in it?

Anyway, as far as I can tell this `continue` should be fine.